### ISSUE TITLE
Dataset created with TarWriter is too large

### ISSUE THREAD

**Issue by ethanabrooks** [webdataset/webdataset#261]
I am creating a dataset with the following code:
```python
import webdataset as wds
from tqdm import tqdm
import io
import torch
from tensordict import TensorDict

N = 20_000
inps = torch.rand(N, 30)
tgts =  torch.arange(N, dtype=torch.float32)
tensordict = TensorDict(dict(inps=inps, tgts=tgts), batch_size=[N])

with wds.TarWriter(f"/tmp/dest.tar") as sink:
    for i, d in tqdm(enumerate(tensordict), total=N):
        obj = {"__key__": f"{i}"}
        for k, v in d.items():
            buffer = io.BytesIO()
            torch.save(v, buffer)
            obj[f"{k}.pth"] = buffer.getvalue()
        sink.write(obj)
```
The resulting tar file is 47Gb:
```
❯ ls -lah /tmp/dest.tar
-rw-rw-r-- 1 ethanbro ethanbro 47G Apr  8 12:25 /tmp/dest.tar
```
I don't understand why the file is so large. If I write the full tensors, without breaking them up by row, the file is a fraction of the size:
```python
with wds.TarWriter(f"/tmp/dest.tar") as sink:
    obj = {"__key__": "key"}
    for k, v in tensordict.items():
        buffer = io.BytesIO()
        torch.save(v, buffer)
        print(v.shape)
        obj[f"{k}.pth"] = buffer.getvalue()
    sink.write(obj)
```
```
❯ ls -lah /tmp/dest.tar
-rw-rw-r-- 1 ethanbro ethanbro 2.4M Apr  8 12:40
```
With this smaller file, I lose a lot of the desired functionality of webdataset. For example, shuffling the datapoints becomes cumbersome. What is the recommended approach?

**Comment by tmbdev** [comment]
This has nothing to do with WebDatset. Rather, the vector "v" that tensordict returns to you points to some huge underlying byte array buffer. So, instead of 120 bytes to save your 30 floating point numbers, you are saving 240000 bytes.

A good way of debugging this sort of thing is to look at the tar file itself.

Here is a notebook explaining the problem:

https://colab.research.google.com/drive/1sl-73QpfzycTYDUXCDD19chexzRb8hs_?usp=sharing

However, saving many small tensors still has significant overhead. To avoid this, you can save them in batches and then unbatch before reading.

```
ds = WebDataset("batched-data.tar").decode(...).unbatched(). etc.
```

You can also batch and unbatch the compressed data, you can save complete batches as `.mp` files, etc.
